substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/PreMainSupport.java

```diff
@@ -0,0 +1,231 @@
+ *           // JVM runtime
+ *         }
+ *         </pre>
+ *     </code></li>
+ * <li>Use {@link com.oracle.svm.core.annotate.TargetClass} API to implement a native image version
+ * premain.</li>
+ * </ul>
+ * </p>
+ */
+public class PreMainSupport {
+
+    // todo: Not finally decided, could be changed in the future
+    private static final String PREMAIN_OPTION_PREFIX = "-XX-premain:";
+
+    /**
+     * A record for premain method.
+     *
+     * @param className the full qualified class name that declares the premain method
+     * @param method the premain method
+     * @param args the arguments of premain method
+     */
+    record PremainMethod(String className, Method method, Object[] args) {
+    }
+
+    private Map<String, String> premainOptions = new HashMap<>();
+    // Order matters
+    private List<PremainMethod> premainMethods = new ArrayList<>();
+
+    @Platforms({Platform.HOSTED_ONLY.class})
+    public void registerPremainMethod(String className, Method executable, Object... args) {
+        premainMethods.add(new PremainMethod(className, executable, args));
+    }
+
+    /**
+     * Retrieve premain options from input args. Keep premain options and return the rest args as
+     * main args. Multiple agent options should be given in separated {@code -XX-premain:} leading
+     * arguments. The premain options format: <br>
+     * -XX-premain:[full.qualified.premain.class]:[premain options]
+     * -XX-premain:[full.qualified.premain.class2]:[premain options] <br>
+     *
+     * @param args original arguments for premain and main
+     * @return arguments for main class
+     */
+    public String[] retrievePremainArgs(String[] args) {
+        List<String> mainArgs = new ArrayList<>();
+
+        for (String arg : args) {
+            if (arg.startsWith(PREMAIN_OPTION_PREFIX)) {
+                String premainOptionKeyValue = arg.substring(PREMAIN_OPTION_PREFIX.length());
+                String[] pair = premainOptionKeyValue.split(":");
+                if (pair.length == 2) {
+                    premainOptions.put(pair[0], pair[1]);
+                }
+            } else {
+                mainArgs.add(arg);
+            }
+        }
+        return mainArgs.toArray(new String[0]);
+    }
+
+    public void invokePremain() {
+        for (PremainMethod premainMethod : premainMethods) {
+            try {
+                Object[] args = premainMethod.args;
+                if (premainOptions.containsKey(premainMethod.className)) {
+                    args[0] = premainOptions.get(premainMethod.className);
+                }
+                // premain method must be static
+                premainMethod.method.invoke(null, args);
+            } catch (Throwable t) {
+                VMError.shouldNotReachHere("Fail to execute " + premainMethod.className + ".premain", t);
+            }
+        }
+    }
+
+    /**
+     * This class is a dummy implementation of {@link Instrumentation} interface. It serves as the
```

Comment:
We need to mention in the release notes that this is experimental and that the behavior of some methods might change. We should test only the methods for which we are completely certain they will stay the same.